### PR TITLE
Enhancements to XLSX and ODS drivers

### PR DESCRIPTION
**Pull request by mccuskk**
This builds on my last patch but includes a number of enhancements. I'm not sure what you will make of them but they build on our actual experience of trying to read client supplied files (which are usually automatically generated by their Housing Management Systems). 

What these changes try to is make the drivers more robust to detecting empty rows and empty cells and provide a couple of configuration options to stop the driver returning millions of empty rows or causing header columns recognition to fail due to empty columns or random whitespace columns in the data beyond the header columns.

I can well see you might see these changes as beyond the scope of GDAL, but if you are interested then let me know and I can add autotests and any cleanups you want.

Summary of changes:

1. ODS Filter: Misdetection of field type if first cell blank   
2. Allow ODS driver to recognise ODFDate fields   
3. Allow empry rows to be optionally ignored in ODS and XLSX drivers using      
    OGR_[XLSX|ODS]_EMPTY_ROWS IGNORE 
4. Allow cells containing only whitespace to be treated as empty in ODS and XLSX drivers using      
    OGR_[XLSX|ODS]_EMPTY_CELLS IGNORE   
5. Fix XML date parsing to not give 12:30:60 (rounding issue)

    
Many Thanks

Kieran

### CODE REVIEW

File: gdal/ogr/ogrutils.cpp [hunk 1]
```diff
@@ -1223,13 +1223,23 @@ int OGRParseXMLDateTime( const char* pszXMLDateTime,
              (c == '+' || c == '-') )
     {
         TZ = 100 + ((c == '+') ? 1 : -1) * ((TZHour * 60 + TZMinute) / 15);
+        if( second >= 59.99)
```
Comment (rouault) [inline gdal/ogr/ogrutils.cpp]: When is this change really necessary ? There is an edge case where it is not correct: leap seconds where second = 60.x is valid

**Comment by rouault** [comment]
- Would you mind rebasing on top of latest trunk ?
- Documenting the new options in the HTML files close to the driver code would be needed, as well as tests
- I don't understand the need for the XML date time parsing change

**Comment by mccuskk** [comment]
Hi

I am just rebasing and adding autotests but would like your thoughts on an issue I have encountered. I have looked again at my change for XML date time parsing change and realised I hadn't looked carefully enough. I think I have now and there seems to be a issue around milliseconds. The current code says if the milliseconds are less than 0.0005 then treat them as 0, what the code doesn't say is what to do with milliseconds like 0.999998 where it ends up formatting them as %06.3f which gives me my 2017/12/29 15:49:60.000 (rounding by vsnprintf). What I have done is change the vsnprintf formatting string to %09.6f which works as I then get 2017/12/29 15:49:59.999996. 

Is this what you would want? I will have to adjust the tests as setting a field to '2015/12/23 12:34:56.789' and querying it's value gives '2015/12/23 12:34:56.789001'.

Alternately I could test if the milliseconds are 995 or greater (similar logic to the 0.005) and round the datetime up by a second. Writing this has made me think this is the best approach - What do you think?

Thanks

Kieran

**Comment by rouault** [comment]
> Alternately I could test if the milliseconds are 995 or greater (similar logic to the 0.005) and round the datetime up by a second. Writing this has made me think this is the best approach - What do you think?

For the sake of simplicity, I'd just convert ms = 1000 to ms = 999.

**Comment by mccuskk** [comment]
I've actually coded up the 995 millisecond rounding in my latest changes. It has the advantage that you see the value that Excel and Calc show you in the spreadsheet.

**Comment by mccuskk** [comment]
I think it's good to go after the last fixes I pushed up, subject to any changes you would like of course.

Thanks

Kieran

**Comment by rouault** [comment]
This fails to build on Windows: https://ci.appveyor.com/project/OSGeo/gdal/build/1.0.4430/job/avw7qwunmwtojwkw
I'd really prefer if we could remove all the mktime() + gmtime() complication, and just set to ms=999 when ms=1000 to avoid changing second

**Comment by mccuskk** [comment]
Ok no bother. I'll rework the patch. Just note that this will show .995
whereas excel and calc will show the next rounded date time.

Kieran

**Comment by mccuskk** [comment]
I've just committed a fixed version of rounded date time for the windows build just so I have a working version. Once it passes checks I'll replace it with a version with the ms changed to 999 version.

File: gdal/ogr/ogrsf_frmts/xlsx/ogrxlsxdatasource.cpp [hunk 2]
```diff
@@ -639,25 +647,13 @@ void OGRXLSXDataSource::startElementTable(const char *pszNameIn,
             return;
         }
         nNewCurLine --;
-        const int nFields = std::max(
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/xlsx/ogrxlsxdatasource.cpp]: Why has been this logic removed ? This helps fixing various timeouts / out of memory situation on fuzzed files

File: gdal/ogr/ogrsf_frmts/ods/ogrodsdatasource.cpp [hunk 3]
```diff
@@ -657,29 +665,33 @@ void OGRODSDataSource::startElementTable(const char *pszNameIn,
...
-            return;
+          nRowsRepeated = 1;
         }
-        const int nFields = std::max(
-            static_cast<int>(apoFirstLineValues.size()),
-            poCurLayer != nullptr ?
-                poCurLayer->GetLayerDefn()->GetFieldCount() : 0);
-        if( nFields > 0 && nRowsRepeated > 100000 / nFields )
+        
+        nRowsRepeated = std::min(nRowsRepeated, 10000);
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/ods/ogrodsdatasource.cpp]: This will silently clamp number-rows-repeated. I prefered the previous explicit error strategy

File: gdal/ogr/ogrfeature.cpp [hunk 4]
```diff
@@ -2036,14 +2036,54 @@ static void OGRFeatureFormatDateTimeBuffer( char szTempBuffer[TEMP_BUFFER_SIZE],
...
+            struct tm t;
+            t.tm_year = nYear - 1900;
+            t.tm_mon = nMonth - 1;
+            t.tm_mday = nDay;
+            t.tm_hour = nHour;
+            t.tm_min = nMinute;
+            t.tm_sec = static_cast<int>(fSecond) + 1;
+            t.tm_isdst = 0;
+            
+            time_t when = mktime(&t);
```
Comment (rouault) [inline gdal/ogr/ogrfeature.cpp]: Instead of mktime() / localtime(), you could probably use the CPLYMDHMSToUnixTime() / CPLUnixTimeToYMDHMS() functions from cpl_time.h that don't suffer the tread-safety issue of localtime()

File: gdal/ogr/ogrfeature.cpp [hunk 5]
```diff
@@ -2036,14 +2036,54 @@ static void OGRFeatureFormatDateTimeBuffer( char szTempBuffer[TEMP_BUFFER_SIZE],
...
-        CPLsnprintf( szTempBuffer, TEMP_BUFFER_SIZE,
-                  "%04d/%02d/%02d %02d:%02d:%06.3f",
-                  nYear,
-                  nMonth,
-                  nDay,
-                  nHour,
-                  nMinute,
-                  fSecond );
+    {
+        if ( ms == 1000 )
```
Comment (rouault) [inline gdal/ogr/ogrfeature.cpp]: This logic for dealing with ms == 1000 is repeated 3 times. It would be good to have it in a single function called from those 3 call sites

**Comment by rouault** [comment]
Due to the migration to github and history rewriting this PR has been automatically closed, please recreate your git clone and issue a new pull request
